server/src/main/java/org/jboss/as/server/controller/git/ElytronClientSshdSessionFactory.java

```diff
@@ -0,0 +1,237 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
+import java.security.AccessController;
+import java.security.GeneralSecurityException;
+import java.security.KeyPair;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+
+import org.eclipse.jgit.annotations.NonNull;
+import org.eclipse.jgit.errors.TransportException;
+import org.eclipse.jgit.errors.UnsupportedCredentialItem;
+import org.eclipse.jgit.internal.transport.sshd.CachingKeyPairProvider;
+import org.eclipse.jgit.transport.CredentialItem;
+import org.eclipse.jgit.transport.CredentialsProvider;
+import org.eclipse.jgit.transport.URIish;
+import org.eclipse.jgit.transport.sshd.IdentityPasswordProvider;
+import org.eclipse.jgit.transport.sshd.KeyPasswordProvider;
+import org.eclipse.jgit.transport.sshd.SshdSession;
+import org.eclipse.jgit.transport.sshd.SshdSessionFactory;
+import org.eclipse.jgit.util.FS;
+import org.jboss.as.server.logging.ServerLogger;
+import org.wildfly.client.config.ConfigXMLParseException;
+import org.wildfly.security.auth.callback.CredentialCallback;
+import org.wildfly.security.auth.client.AuthenticationConfiguration;
+import org.wildfly.security.auth.client.AuthenticationContext;
+import org.wildfly.security.auth.client.AuthenticationContextConfigurationClient;
+import org.wildfly.security.auth.client.ElytronXmlParser;
+import org.wildfly.security.credential.KeyPairCredential;
+import org.wildfly.security.credential.PasswordCredential;
+import org.wildfly.security.credential.SSHCredential;
+import org.wildfly.security.password.interfaces.ClearPassword;
+
+
+/**
+ * SshdSessionFactory for JGIt using Elytron.
+ * Supports specifying ssh-keys as {@link KeyPairCredential} or {@link SSHCredential}
+ *
+ * @author <a href="mailto:aabdelsa@redhat.com">Ashley Abdel-Sayed</a>
+ */
+public class ElytronClientSshdSessionFactory extends SshdSessionFactory {
+
+    private static AuthenticationContextConfigurationClient CLIENT = AccessController.doPrivileged(AuthenticationContextConfigurationClient.ACTION);
+    private final AuthenticationContext context;
+
+    private final Map<Tuple, Iterable<KeyPair>> defaultKeys;
+    private String[] defaultIdentities;
+    private KeyPair keyPair;
+    private URI uri;
+    private String knownHostsFile;
+
+    public ElytronClientSshdSessionFactory(URI authenticationConfig) throws ConfigXMLParseException, GeneralSecurityException {
+        if (authenticationConfig != null) {
+            context = ElytronXmlParser.parseAuthenticationClientConfiguration(authenticationConfig).create();
+        } else {
+            context = null;
+        }
+
+        this.defaultKeys = new ConcurrentHashMap();
```

Comment:
This should be `new ConcurrentHashMap<>()` so we don't get a compile warning.

Reply:
Fixed, thanks!